src/chain.h

```diff
@@ -126,7 +126,15 @@ enum BlockStatus: uint32_t {
     BLOCK_FAILED_CHILD       =   64, //!< descends from failed block
     BLOCK_FAILED_MASK        =   BLOCK_FAILED_VALID | BLOCK_FAILED_CHILD,
 
-    BLOCK_OPT_WITNESS       =   128, //!< block data in blk*.data was received with a witness-enforcing client
+    BLOCK_OPT_WITNESS        =   128, //!< block data in blk*.data was received with a witness-enforcing client
+
+    /**
+     * If set, this indicates that the block index entry is assumed-valid.
+     * Certain diagnostics will be skipped in e.g. CheckBlockIndex().
+     * It almost certainly means that the block's full validation is pending
+     * on a background chainstate. See `docs/assumeutxo.md`.
+     */
+    BLOCK_ASSUMED_VALID      =   256,
```

Comment:
I low-key agree.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/21526#discussion_r696190873

James asked me about the flag again, and I don't have much new to add so I'll just restate what I wrote https://github.com/bitcoin/bitcoin/pull/21526#discussion_r674198521 and https://github.com/bitcoin/bitcoin/pull/21526#discussion_r683813701, which is that I don't like complexity of flags being added and then removed instead of just being cumulative. I don't like flags trying to tell me what I should be assuming about a block, instead of just telling me what kind of validation was actually done on the block. So I'd drop the "assume" flag and use something like BLOCK_VALIDATED_FROM_GENESIS, BLOCK_VALIDATED_FROM_SNAPSHOT flags instead. Or not have any new flag and just have validated begin/end pointers or height ranges.

But assuming flag implementation isn't going to change at this point, ariard's suggestion to add something about utxos or snapshots to the flag name does not seem like a bad idea. I don't really see how confusion with the assumevalid feature would be possible, but also don't see drawbacks to making the flag name more descriptive.

And all of this is definitely bikeshedding! Current flag seems perfectly reasonable and not confusing in any way I could realistically see.

Reply:
Thanks again @ryanofsky for weighing in.

> I don't like complexity of flags being added and then removed instead of just being cumulative

I agree that ideally it would be better to set flags once and never touch them subsequently, but there is a need to remove the ASSUMED flag on block index entries as they're validated so that CheckBlockIndex eventually returns to its normal behavior. I'm sure there's some way we could do that with some combination of flags that are set once and never unset, but I think that would probably require a much bigger change and is more trouble than I think is worth.

---

@ariard @naumenkogs unless there is significant dissent, I'm keeping this name as-is. Part of what motivated this change in the first place was to avoid leaking "assumeutxo" specifics out into the rest of the system; if we change the name of this flag, we then have to change `pindex->IsAssumedValid()` to something like `IsAssumedUtxo()` which then leaks that detail again; and I don't find that name change to be sufficiently descriptive to make that worthwhile.

BLOCK_ASSUMED_VALID is the correct name here because the block is assumed to be valid; it has not yet actually been encountered by the system. I don't think that confusion with the `-assumevalid=` CLI flag (which should really be something like `-skipsigsbeneath=`) is really going to trip someone up here looking at block index flags. The other thing is that BLOCK_UTXO_ASSUMED doesn't strike me as sufficiently descriptive - the block index shouldn't care _why_ we're assuming the thing valid, and if we end up doing parallel validation with something like utreexo, the over-specified name will only be more confusing and out of date.

Reply:
Not gonna insist.
Perhaps, the confusion with `-assumevalid=` matters only for me, someone who looks at this part of the code for the second time.

Reply:
One of the benefits of the 'cumulative' flags is that you can prove the block validity state transitions are finite through consumption of the flag.

However, it's easy to prove this by inspection as well, so I think the current design is OK.

Reply:
>  So I'd drop the "assume" flag and use something like BLOCK_VALIDATED_FROM_GENESIS, BLOCK_VALIDATED_FROM_SNAPSHOT flags instead. 

Yes, I think flags informing on the level of validation which has been performed so far would be better than tagging blocks with the validation process they belong to.

I agree, let's not stuck on that, we can improve with 'cumulative' flags and better description of the different validation models (e.g `skipsigbeneath`) in the future.